src/GovUk.Education.ExploreEducationStatistics.Data.Model/Migrations/Current_Routine_RebuildIndexes.sql

```diff
@@ -1,120 +1,302 @@
+                                                  ObjectName, StartFragPercent, ActionRequired)
+    SELECT Id,
+           @RunId,
+           IndexName,
+           SchemaName,
+           ObjectName,
+           FragPercent,
+           ActionRequired
+    FROM #Stats;
+
+    -- Create rows for tables
+    INSERT INTO __Log_RebuildIndexesUpdateStatistics (RunId, TableName)
+    SELECT @RunId,
+           TL.ObjectName
+    FROM #TableList TL;
+
+    /*-----------------------------------------------------------------------*/
+
+    -- Rebuild or reorganize each index depending on the fragmentation percentage
+    SET @StatsRowIndex = 1;
+
+    WHILE @StatsRowIndex <= @StatsCount AND @SkipRemainingTasks = 0
+        BEGIN
+            SELECT @Command =
+                   'ALTER INDEX ' + IndexName + ' ON ' + SchemaName + '.' + ObjectName +
+                   IIF(ActionRequired = 'Rebuild',
+                       ' REBUILD WITH (ONLINE = ON)',
+                       ' REORGANIZE'),
+                   @FragPercentBefore = FragPercent,
+                   @IndexName = IndexName,
+                   @SchemaName = SchemaName,
+                   @ObjectName = ObjectName,
+                   @ActionRequired = ActionRequired
+            FROM #Stats
+            WHERE Id = @StatsRowIndex;
+
+            -- Save start time for reporting
+            UPDATE __Log_RebuildIndexesAlterIndexes
+            SET StartTime = GETUTCDATE()
+            WHERE RunId = @RunId
+              AND ObjectName = @ObjectName
+              AND SchemaName = @SchemaName
+              AND IndexName = @IndexName;
+
+            -- Do this check after setting StartTime, as we rely on StartTime to determine which indexes we're going
+            -- to skip if we pass @StopTime
+            IF @ActionRequired = 'None'
+                BEGIN
+                    RAISERROR ('Skipping command - fragmentation percent did not hit reorganisation threshold ''%s''.', 0, 1,
+                        @Command) WITH NOWAIT;
                     SET @StatsRowIndex += 1;
+                    CONTINUE;
                 END
-            IF(@StopTime > GETUTCDATE())
-                RAISERROR ('Completed optimising indexes for tables.', 0, 1, NULL) WITH NOWAIT;
 
-            /*-----------------------------------------------------------------------*/
+            -- Reorganise or rebuild
+            RAISERROR ('Executing command ''%s''.', 0, 1, @Command) WITH NOWAIT;
+            EXEC (@Command);
 
-            SET @StatsRowIndex = 1;
-            SET @LastTable = '';
-            SELECT @StatsCount = COUNT(Id) FROM #Stats;
-            WHILE(@StatsRowIndex <= @StatsCount AND @StopTime > GETUTCDATE())
+            -- Save end info for reporting
+            UPDATE __Log_RebuildIndexesAlterIndexes
+            SET __Log_RebuildIndexesAlterIndexes.EndTime        = GETUTCDATE(),
+                __Log_RebuildIndexesAlterIndexes.EndFragPercent = Fragmented.FragPercent
```

Comment:
I think I still have performance concerns about running this for a second time albeit on an object by object basis, but I guess we could see how it goes. Have you tried running this query for just one index on one of the big tables to see how long this takes?

`EndTime` is going to be recorded as the end time plus whatever time this takes, rather than the time `EXEC (@Command)` finished executing so could be a bit misleading if this takes a long time.

Reply:
I tested this against the prod database - every index for the Observation and ObservationFilterItems tables took 8 seconds to fetch the fragmentation percent, except for IX_Observation_LocationId, which took 5 mins 30 secs.

Discussed this with Ben and we're going to leave this for now. We can work out how long the initial fetching of all frag percents takes by comparing the overall start time with the start time of the first index. Once we have that, we can determine whether we want to do other changes here, such as ensuring fetching the EndFragPercent doesn't increase the Start/End times we record of the reindexing or even the more extreme case of dropping EndFragPercent entirely.